### PR TITLE
Rewrite simple filters for bounded `integers()`

### PR DESCRIPTION
**Pull request by Zac-HD**
This pull request is the first piece of #2701.

It demonstrates that the concept works, handles the tricky interactions with LazyStrategy and chained filters (though `one_of()` is left for future work), and provides obvious locations for future contributions - such as copying this logic for `floats()`.  Further discussion of how the overall plan can be broken down into pull requests can be found at #2701, along with next steps.

### CODE REVIEW

File: hypothesis-python/src/hypothesis/internal/filtering.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    # TODO: handle lambdas by AST analysis
+
+    return {}, predicate
+
+
+def get_integer_predicate_bounds(predicate: Predicate) -> ConstructivePredicate:
+    kwargs, predicate = get_numeric_predicate_bounds(predicate)
+
+    if "min_value" in kwargs:
+        print(math.isinf(kwargs["min_value"]))
```
Comment (rsokl) [inline hypothesis-python/src/hypothesis/internal/filtering.py:98]: I'm guessing this is a straggler?
```suggestion
```

Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/filtering.py:98]: Whoops, yeah.

File: hypothesis-python/src/hypothesis/strategies/_internal/strategies.py [hunk 2]
```diff
@@ -737,15 +736,34 @@ def __repr__(self):
...
+            FilteredStrategy.__init__(self, fresh, ())
+
+    def filter(self, condition):
+        # Allow strategy rewriting to 'see through' an unhandled predicate.
+        out = self.filtered_strategy.filter(condition)
+        if isinstance(out, FilteredStrategy):
+            return FilteredStrategy(
+                out.filtered_strategy, self.flat_conditions + out.flat_conditions
+            )
+        return FilteredStrategy(out, self.flat_conditions)
```
Comment (rsokl) [inline hypothesis-python/src/hypothesis/strategies/_internal/strategies.py:756]: Please add comments here. The flow here is hard to follow, especially since the first thing that `.filter()` does is call `.filter()`..

File: hypothesis-python/tests/cover/test_filtered_strategy.py [hunk 3]
```diff
@@ -25,3 +26,22 @@ def test_filter_iterations_are_marked_as_discarded():
...
+    s = FilteredStrategy(st.integers() | st.text(), (bool,))
+    assert all(isinstance(x, FilteredStrategy) for x in s.branches)
+
+
+def test_filter_conditions_may_be_empty():
+    s = FilteredStrategy(st.integers(), conditions=())
+    s.condition(0)
+
+
+def test_flattens_nested_filteredstrategy():
```
Comment (rsokl) [inline hypothesis-python/tests/cover/test_filtered_strategy.py:41]:
```suggestion
def test_nested_filteredstrategy_flattens_conditions():
```

**Review by rsokl — CHANGES_REQUESTED** [review]
Looks good! Just some minor requests

File: hypothesis-python/src/hypothesis/internal/filtering.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+def get_integer_predicate_bounds(predicate: Predicate) -> ConstructivePredicate:
+    kwargs, predicate = get_numeric_predicate_bounds(predicate)
+
+    if "min_value" in kwargs:
+        if kwargs["min_value"] == -math.inf:
+            del kwargs["min_value"]
+        elif math.isinf(kwargs["min_value"]):
+            return {}, lambda _: False
+        elif kwargs["min_value"] != int(kwargs["min_value"]):
```
[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/filtering.py:102]: `isclose` would be a change of behaviour, and thus unacceptable in this context... it's unfortunate if someone expected `integers().filter(lambda x: x >= epsilon)` to allow `0`, but since `0 < epsilon` there's really not much else to say.

[the remark this answers is not in the document]
Comment (rsokl) [reply hypothesis-python/src/hypothesis/internal/filtering.py:113]: Ah, of course. You're totally right. I wasn't thinking about this correctly

**Review by rsokl — APPROVED** [review]
